File: src/exercise2/java/SwindleConcurrent.java
import java.time.Month;
import java.util.List;

public class SwindleConcurrent {
    public static void main(String[] args) {
        Backend backend = new Backend();
        int userId = 1;

        List<Thread> threads = List.of(
            new Thread(() -> backend.onAddClick(userId, Month.JULY)),
            new Thread(() -> backend.onAddClick(userId, Month.JULY)),
            new Thread(() -> backend.onAddClick(userId, Month.JULY)),
            new Thread(() -> backend.onAddClick(userId, Month.JULY)),
            new Thread(() -> backend.onAddClick(userId, Month.JULY)),
            new Thread(() -> backend.onAddClick(userId, Month.AUGUST)),
            new Thread(() -> backend.onAddClick(userId, Month.AUGUST)),
            new Thread(() -> backend.onAddClick(userId, Month.AUGUST)),
            new Thread(() -> backend.onAddClick(userId, Month.AUGUST))
        );
        threads.forEach((thread) -> {
            thread.start();
            try {
                Thread.sleep(30);
            } catch (InterruptedException e) { }
        });

        threads.forEach((thread) -> {
            try {
                thread.join();
            } catch (InterruptedException e) { }
        });

        backend.cashOut(userId, Month.JULY);
        backend.cashOut(userId, Month.AUGUST);
    }
}

File: src/exercise2/go/swindleConcurrent.go
package main

import (
	"sync"
	"time"
)

func main() {
	backend := Backend{NewDatabase()}
	userId := 1
	var wg sync.WaitGroup
	wg.Add(9)

	go backend.onAddClickConcurrent(userId, time.July, &wg)
	time.Sleep(30 * time.Millisecond)
	go backend.onAddClickConcurrent(userId, time.July, &wg)
	time.Sleep(30 * time.Millisecond)
	go backend.onAddClickConcurrent(userId, time.July, &wg)
	time.Sleep(30 * time.Millisecond)
	go backend.onAddClickConcurrent(userId, time.July, &wg)
	time.Sleep(30 * time.Millisecond)
	go backend.onAddClickConcurrent(userId, time.August, &wg)
	time.Sleep(30 * time.Millisecond)
	go backend.onAddClickConcurrent(userId, time.August, &wg)
	time.Sleep(30 * time.Millisecond)
	go backend.onAddClickConcurrent(userId, time.August, &wg)
	time.Sleep(30 * time.Millisecond)
	go backend.onAddClickConcurrent(userId, time.August, &wg)
	time.Sleep(30 * time.Millisecond)
	go backend.onAddClickConcurrent(userId, time.August, &wg)

	wg.Wait()

	backend.cashOut(userId, time.July)
	backend.cashOut(userId, time.August)
}
